args/core/async/readonly_rw_spinlock.hpp

```diff
@@ -161,26 +395,137 @@ namespace args::core::async
 		 */
 		~readwrite_guard()
 		{
-			if (preacquired) // Another write guard is still alive that will unlock the lock for this thread.
-				return;
+			m_lock.unlock(write);
+		}
+
+		readwrite_guard& operator=(readwrite_guard&&) = delete;
+	};
+
+
+	/**@class readwrite_multiguard
+	 * @brief RAII guard that uses multiple ::async::readonly_rw_spinlocks to lock them all for read-write. (similar to std::lock)
+	 * @note Read-Write operations cannot happen simultaneously and will wait for each other.
+	 *		 Read-Write operations will also wait for any Read-only operations to be finished.
+	 * @ref args::core::async::readonly_rw_spinlock
+	 */
+	class readwrite_multiguard
+	{
+	private:
+		std::vector<readonly_rw_spinlock*> m_locks;
+
+	public:
+		/**@brief Creates read-write multi-guard and locks for Read-Write.
+		 */
+		template<typename lock_type1, typename lock_type2, typename... lock_typesN>
+		readwrite_multiguard(lock_type1& lock1, lock_type2& lock2, lock_typesN&... locks)
+		{
+			m_locks.push_back(&lock1);
+			m_locks.push_back(&lock2);
+			(m_locks.push_back(&locks), ...);
+
+			sparse_set<uint> unlockedLocks;
+			for (uint i = 0; i < m_locks.size(); i++)
+				unlockedLocks.insert(i);
 
-			if (returnToRead) // read permission was granted before write request, we should return to read instead of idle after write is finished.
+			bool locked = true;
+			do
 			{
-				// We should be the only one to have had access so we can safely write to the lock state and readers.
-				lock.readers.fetch_add(1, std::memory_order_relaxed);
-				lock.readState.store(readonly_rw_spinlock::read, std::memory_order_release);
+				locked = true;
+				for (uint i : unlockedLocks)
+				{
+					if (m_locks[i]->try_lock(write))
+						unlockedLocks.erase(i);
+					else
+						locked = false;
+				}
+			} while (!locked);
```

Comment:
how does this actually prevent deadlocks ? shouldn't all locks be unlocked on an failed attempt in order to not spin endlessly ?

Reply:
yes you're right...